### DISCUSSION TITLE
에디터에 ZepetoScript 프로퍼티 노출은 지원되지 않나요?

### DISCUSSION THREAD

**Discussion by Cha-SeongHun** [naverz/zepeto-studio-kor#504] [5. Q&A]
아래와 같은 제페토스크립트가 있을 때, Transform은 노출되지만 ZepetoScript는 프로퍼티로 노출되지 않습니다.
```ts
export default class Test extends ZepetoScriptBehaviour {
  public testTransform: Transform;
  public testScript: ScriptA;
}
```

의도된 사양인지 혹은 지원될 예정인지 궁금합니다

**Comment by Zepeto-TechEvangelist** [comment]
안녕하세요. 제페토 스크립트간 연결은 아래의 내용을 참고 부탁 드립니다.
https://studio.zepeto.me/kr/guides/script-import

**Reply by TaejunPark** [reply]
https://github.com/naverz/zepeto-studio-kor/discussions/463#discussion-3920317
위에서처럼,  GetComponent<xxx>로 typescript 불러올 때  TypeError: Cannot read property 'GetComponent' of undefined 로 안되고, 
프로퍼티로도 활용이 안되니까, 씬 안의 오브젝트에 있는 제페토스크립트의 메소드를 활용할 일반적인 방법이 당장 잘 안떠오릅니다..
뭔가 방법이 있을까요?


처음에 다른 타입스크립트 오브젝트에서 Starter 오브젝트 메소드를 쓰려고 GetCompoent<>() 실행해서 호출했을 땐 돼서 잘 하고 있었는데,
하루이틀? 작업하고 있다가 동일한 행동을 Starter에서 하려니까 안됩니다.
뭔가 캐시?가 잘못된거라서  뭘 리셋하면 된다든가 하는 해결책이 있을까여?


아마 작성자분도 이래 저래 해보신 것 같은데, 저도 비슷한 경과로 따라한 것 같습니다.. 질문 올리기 싫었는데
아무 의미없는 것들 알아내는 데에 계속 시간이 많이 쓰이니 참 슬프네요..

**Reply by Zepeto-TechEvangelist** [reply]
원하는 메소드가 들어있는 제페토 스크립트를 싱글톤으로 선언해서 사용하는 방법이 있습니다.
아래는 예제입니다.

UIController를 싱글톤으로 선언하고, 아이템을 플레이어가 먹었을때 카운트를 올리는 상황입니다.

### [ 싱글톤 선언 ]

<img width="738" alt="스크린샷 2022-03-25 오후 4 09 07" src="https://user-images.githubusercontent.com/100118650/160071991-ddcdf657-d7fe-4d9b-bceb-aa8d6109d419.png">

### [ UIController에 있는 IncreaseCount 메소드 호출 ]

<img width="665" alt="스크린샷 2022-03-25 오후 4 10 19" src="https://user-images.githubusercontent.com/100118650/160072177-99fcfecc-230a-4f62-82b1-7e34dee4617f.png">

**Reply by TaejunPark** [reply]
엄청 빠른 답변 감사합니다,

잠깐 또 보다보니까 뭐 좀 해봤는데 되네요..
저처럼 타입스크립트인지 아니면 뭔가에 안 익숙한 분들은 틀릴 수 있을 것 같으니 나중에 어디 적용된 예제가 있다면 좋을 것 같습니다.

함수를 부를 때, 리더보드 예제에서
 LeaderboardAPI.SetScore(this.leaderboardId, this.score, this.OnResult, this.OnError);
에서, this.OnResult 처럼 함수로 변수를 넘겨주면
 this.OnResult 함수 내에선, 에디터에서 설정해준 프로퍼티 this.xxx 는 새로운 변수?로 접근이 되는 것 같습니다. 그래서 프로퍼티가 undefined로 되는 것 같습니다. 왠지는 몰라도 함수 내부에서는 계속 값이 이어지는 것 같아서 다음에 부르면 그값이 여전히 있는 것 같습니다.

서버 메시지 받을 때도 비슷한 일이 있었어서 
 room.AddMessageHandler("xxx", this.OnReceivedXXX); 
로 했다가 변수가 계속 undefined로 돼서 한참 헤메다가
 room.AddMessageHandler("xxx", ()=>{}); 
로 바꿔서 썼엇네요..

적어주신 내용 잘 읽어봤습니다, 평소 코딩할때 잘 응용해보도록 하겠습니다.
